Add template for 404 page

EltonGohJH opened the pull request:
**What is the purpose of this pull request?**

- [x] Documentation update
- [ ] Bug fix
- [ ] Feature addition or enhancement
- [ ] Code maintenance
- [ ] DevOps
- [x] Improve developer experience
- [ ] Others, please explain:

<!--
  If this pull request is addressing an issue, link to the issue: "Fixes #xxx" or "Resolves #xxx"

  If the pull request completely addresses the issue, use one of the issue closing keywords.
  https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

  Otherwise, elaborate further on the rationale of this pull request as needed
-->

**Overview of changes:**
Resolves https://github.com/MarkBind/markbind/issues/1899

**Anything you'd like to highlight/discuss:**
Currently, what I have created is very minimalistic. Any inputs on what can should be added will be helpful.

**Testing instructions:**

<!--
  Any special testing instructions **not including** our automated tests.
-->

**Proposed commit message: Add template for 404 page**

<!--
  See this link for more info on how to write a good commit message:
  https://oss-generic.github.io/process/docs/FormatsAndConventions.html#commit-message

  As best as possible, write a succinct commit title in 50 characters

|---------This is the width of 50 chars----------|
|-----------This is the width of 72 chars for your reference-----------|
-->

---

**Checklist:** :ballot_box_with_check:

<!-- Leave non-applicable items unchecked -->

- [x] Updated the documentation for feature additions and enhancements
- [x] Added tests for bug fixes or features
- [x] Linked all related issues
- [x] No unrelated changes <!-- Its tempting, but increases the reviewer's work, and really pollutes the commit history =( -->

<!--
  We'll try our best to get to your PR within a week.
  If we haven't gotten to it then, or if your pull request resolves an urgent item, feel free to give us a ping!
-->

tlylt commented on docs/userGuide/redirectingToCustom404Page.md:
```diff
@@ -0,0 +1,23 @@
+# {{ title }}
+
+<div class="lead" id="overview">
+Most popular static hosting service such as Netlify supports custom routing to a custom 404 html page. A default 404.md page is provided, and you can customise it to create your own 404 page.
```
Perhaps more details/hyperlinks can be included. You can reference the links given in the original issue to see the documentation for those static hosting services such as (e.g. [in Gatsby](https://www.gatsbyjs.com/docs/how-to/adding-common-features/add-404-page/), [Hugo](https://gohugo.io/templates/404/), [Jekyll](https://jekyllrb.com/tutorials/custom-404-page/), [Netlify](https://docs.netlify.com/routing/redirects/redirect-options/#custom-404-page-handling)) 

Some nits:

```suggestion
Most popular static hosting service such as Netlify supports custom routing to a custom 404 HTML page. A default `404.md` file is provided, and you can customize it to create your own 404 page.
```

jovyntls replied:
```suggestion
Many popular static hosting services, such as Netlify, support custom routing to a custom 404 html page. A default 404.md page is provided, and you can customise it to create your own 404 page.
```

tlylt commented on packages/core/template/default/404.md:
```diff
@@ -0,0 +1,4 @@
+<br>
+
+# 404 page not found
+You can edit this file to create your own 404 page.
```
Perhaps as a default, we can provide a "go back"/ “home" button.
As for the note on editing, perhaps can be more specific and say "edit the 404.md file" to customize the 404 page?

tlylt commented:
Have you tested it and are there any screenshots of the 404 page? Would be great to have a preview and it can be included in the "Testing instructions:" section of the PR description

jovyntls commented on docs/userGuide/workingWithSites.md:
```diff
@@ -14,7 +14,8 @@
   ['Deploying the Site', 'deployingTheSite'],
-  ['MarkBind in the Project Workflow', 'markBindInTheProjectWorkflow']
+  ['MarkBind in the Project Workflow', 'markBindInTheProjectWorkflow'],
+  ['Redirecting to Custom 404 Page', 'redirectingToCustom404Page']
```
```suggestion
  ['Redirecting to a Custom 404 Page', 'redirectingToCustom404Page']
```

jovyntls reviewed (commented):
Minor grammatical nits (also agree with @tlylt's comments above - would be good to have a screenshot an example of a broken link that shows the 404 page!)

EltonGohJH commented:
@tlylt 
I know it is been quite a while. I have yet to update the PR but I updated it locally and deployed it to netlify.
https://dulcet-cocada-93b57f.netlify.app/

This is how it will look like.
<img width="1440" alt="image" src="https://user-images.githubusercontent.com/75515229/187121449-8cc6ef08-09f0-466d-8f77-9eab86329abd.png">

damithc commented:
Would it be better if this 404 page is usable even without any further changes by the author? The current one doesn't meet that criterion because it contains a note meant for the author and not the reader.

tlylt commented:
> Would it be better if this 404 page is usable even without any further changes by the author? The current one doesn't meet that criterion because it contains a note meant for the author and not the reader.

Yup we can probably remove this then (since we have it documented), looks like most other default 404 pages for static web pages don't include it as well. (or minimally, make it less prominent)

> I know it is been quite a while. I have yet to update the PR but I updated it locally and deployed it to netlify.
[dulcet-cocada-93b57f.netlify.app](https://dulcet-cocada-93b57f.netlify.app/)

Thanks @EltonGohJH, just tried browsing it and noticed a few things:
- when going to an unknown page that triggers the 404, the style of the page changes slightly as compared to the home page, wonder if that is something that we can fix?
	- e.g 1: https://dulcet-cocada-93b57f.netlify.app/123
	- e.g 2 (which looks diff from 1) https://dulcet-cocada-93b57f.netlify.app/dir/123
- the page title looks like this:
![image](https://user-images.githubusercontent.com/41845017/187123854-417ec694-b139-42c7-ad7d-42f8689756bf.png)

Perhaps we can update the page title accordingly to show something readable? e.g "Page not found"
(example from github pages)
![image](https://user-images.githubusercontent.com/41845017/187124031-5eaebe86-b652-4c24-80a4-8b2f0553f168.png)

Lastly, wonder if u could include a broken link in ur example site to test whether that way to reaching a 404 page works (just to be sure)

EltonGohJH commented:
@tlylt 
https://dulcet-cocada-93b57f.netlify.app/

I have updated the 404 page. And fixed some of the bugs. Under heading1 at home page, i have added a link as requested for you to try reaching 404 page.

<img width="1440" alt="image" src="https://user-images.githubusercontent.com/75515229/187457143-55eeae71-5af4-4a05-a00e-979c5f557743.png">

tlylt commented on packages/core/template/default/404.md:
```diff
@@ -0,0 +1,14 @@
+    layout: 404.md
+</frontmatter>
+
+<body style="font-family: monospace; background-color: #f1f1f1;">
```
For this file, could we somehow achieve the existing frontend with markdown? Reason being that this is a `.md` file and it could perhaps be easier to edit if it is in markdown if possible, what do you think @EltonGohJH?

Also, a minor nit for the current HTML: the `<body>` tag used here may not be appropriate. See screenshot below (the page ends up having two body tags, which might not be very typical:
![image](https://user-images.githubusercontent.com/41845017/188170463-c05b01c7-ed66-4edc-8507-c0ab90f725d5.png)

EltonGohJH replied:
@tlylt 
I updated the page to make it responsive, you can check Netlify.  I tried to use more of markdown/markbind syntax. Do tell me if that is good enough.
I already have the a simple guide on the UG, is that sufficient.

tlylt replied:
> I already have the a simple guide on the UG, is that sufficient.

Are you referring to the following point I made?
> could we implement this in our UG/DG as well to eat our own dog food?

Sorry if it was unclear, I meant to say --> Can we implement the 404 page logic in our user guide and developer guide documentation, such that if there's a 404, it should show the custom 404 page? On top of trying this out ourselves, it also helps testing since we should be able to directly check the outcome of your changes on our PR preview.

EltonGohJH replied:
@tlylt 
I have updated. You should be able to see the 404 in the preview now.

tlylt replied:
@EltonGohJH https://deploy-preview-1955--markbind-master.netlify.app/devguide/333
Seems to have some style issue when I try to visit the 404 page in DG preview

EltonGohJH replied:
@tlylt 
This is because the html generated uses relative pathing to the css file.
Eg.
`    <link rel="stylesheet" href="markbind/css/bootstrap.min.css">`

Is there a reason why we dun use non relative pathing.

So generate this instead
`    <link rel="stylesheet" href="/markbind/css/bootstrap.min.css">`

ang-zeyu replied:
I don't think there's a specific reason, it seems to have always been the case since we assumed pages are located on the filesystem where they are served too.

ang-zeyu replied:
Made a quick change [here](https://github.com/MarkBind/markbind/pull/2032/files).

See if that works and we can get this part unblocked 🙂

ang-zeyu replied:
@EltonGohJH  In case you missed this, this shouldn't pose any issues now.

tlylt reviewed (commented):
Thanks @EltonGohJH for the follow up, changes are going in the right direction, just few more nits:
- could we implement this in our UG/DG as well to eat our own dog food?
- for mobile display, the 404 is not very centered (because it is too big), could we make some adjustment to ensure responsive design?

![image](https://user-images.githubusercontent.com/41845017/188171157-5d524ca8-3ca3-409b-b95d-652e3f24bbf2.png)

tlylt commented on docs/404.md:
```diff
@@ -0,0 +1,8 @@
+    layout: 404.md
+</frontmatter>
+
+
```
May I check if this blank line is necessary?

EltonGohJH replied:
Not necessary

tlylt replied:
If so could we remove it? What do you think?

EltonGohJH replied:
I will update it. 😢

tlylt commented on packages/cli/test/functional/test_site_convert/test_basic_convert/non_markbind_site/404.md:
```diff
@@ -0,0 +1,14 @@
+    layout: 404.md
+</frontmatter>
+
+<body style="font-family: monospace; background-color: #f1f1f1;">
```
Seems like this is the original template design, could we standardize and fix all test suites as well as their generated HTML?

EltonGohJH replied:
OMG. I forgot bout dis

tlylt commented on packages/cli/test/functional/test_site_convert/test_navigation_convert/non_markbind_site/404.md:
```diff
@@ -0,0 +1,14 @@
+    layout: 404.md
+</frontmatter>
+
+<body style="font-family: monospace; background-color: #f1f1f1;">
```
Same as above

tlylt commented on packages/cli/test/functional/test_site_templates/test_default/expected/404.html:
```diff
@@ -0,0 +1,48 @@
+
+<body>
+  <div id="app">
+    <p>
```
I understand that this is generated code, but it looks a little weird in the placement of the `<p>` tag. Wonder if there's any underlying problem here, what do you think?

EltonGohJH replied:
I have no idea what exactly caused this but this occurred when the layout does not have any parent.

Right now ,layouts/404.md looks like this.
`{{ content }}`

When I added the div. The extra <p> tag is gone.
`<div>{{ content }}</div>`

tlylt replied:
Could you paste the resultant html here? I think if it works fine we can include the `<div>` wrapping as u experimented

EltonGohJH replied:
@tlylt 
If you check the file, with the <div> added. The weird p tag is gone.

tlylt commented:
Thanks, @EltonGohJH for the hard work.

Now that we have the 404 feature implemented for our docs, I think it would be nice to include either a link or a screenshot for users to easily navigate/preview how the default 404 page looks like in our UG ->
![image](https://user-images.githubusercontent.com/41845017/193998051-4f84709d-624f-444a-aa01-bb9249e37833.png)

If not possible to include it in this PR, let's put this up as an issue for the future.

Otherwise LGTM

EltonGohJH commented:
@tlylt 
I have removed that extra lines and updated the docs.

ong6 reviewed (approved):
LGTM. I feel the "Page not found" and "File not found" titles can be made into just "Not Found" though, but I'm also ok with the current wordings.

tlylt commented:
Thanks @EltonGohJH, LGTM. I made a minor update as the 404 link in userguide wasn't working locally without the `.html`.
